### PR TITLE
Fix some clone @default-return handlings

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: glib/tests/clone.rs [hunk 1]
```diff
@@ -457,23 +457,33 @@ fn test_clone_macro_default_return() {
...
-    #[allow(dead_code)]
+    #[derive(PartialEq, Debug)]
     struct Bar {
         x: i32,
     }
 
-    let _closure = clone!(@weak v => @default-return Bar { x: 0 }, move || Bar { x: 1 });
+    let closure = clone!(@weak v => @default-return Bar { x: 0 }, move || Bar { x: 1 });
+    let wrapper = clone!(@weak v => move |call: &(dyn Fn() -> Bar)| call());
+    assert_eq!(wrapper(&closure), Bar { x: 0 });
```
Comment (GuillaumeGomez) [inline glib/tests/clone.rs:467]: This test is failing because it seems like it can upgrade `v` twice, which is very weird...

File: glib/tests/clone.rs [hunk 2]
```diff
@@ -457,23 +457,33 @@ fn test_clone_macro_default_return() {
...
         C { x: i32 },
     }
-    let _closure = clone!(@weak v => @default-return Enum::A, move || Enum::A);
-    let _closure = clone!(@weak v => @default-return Enum::B(0), move || Enum::A);
-    let _closure = clone!(@weak v => @default-return Enum::C { x: 0 }, move || Enum::A);
-    let _closure = clone!(@weak v => @default-return { let x = 12; x + 2 }, move || 19);
+    let closure1 = clone!(@weak v => @default-return Enum::A, move || Enum::B(0));
+    let closure2 = clone!(@weak v => @default-return Enum::B(0), move || Enum::A);
+    let closure3 = clone!(@weak v => @default-return Enum::C { x: 0 }, move || Enum::A);
+    let closure4 = clone!(@weak v => @default-return { let x = 12; x + 2 }, move || 19);
```
Comment (GuillaumeGomez) [inline glib/tests/clone.rs:478]: This test is failing too because it returns 19 instead of 14, no clue why either...

File: glib/tests/clone.rs [hunk 3]
```diff
@@ -457,23 +457,33 @@ fn test_clone_macro_default_return() {
...
+
+    let wrapper = clone!(@weak v => move |call: &(dyn Fn() -> Enum)| call());
+    assert_eq!(wrapper(&closure1), Enum::B(0));
+    assert_eq!(wrapper(&closure2), Enum::A);
+    assert_eq!(wrapper(&closure3), Enum::A);
+
+    let wrapper = clone!(@weak v => move |call: &(dyn Fn() -> usize)| call());
+    assert_eq!(wrapper(&closure4), 14);
```
Comment (GuillaumeGomez) [inline glib/tests/clone.rs:486]: This test is failing too because it returns 19 instead of 14, no clue why either...

File: glib/tests/clone.rs (hunk 1 above)
Comment (sdroege) [reply glib/tests/clone.rs:467]: Can you provide the `cargo expand` output for the failing tests?

Comment (GuillaumeGomez) [reply glib/tests/clone.rs:467]: Enjoy! So this is `closure`:

```rust
{
    let v = glib::clone::Downgrade::downgrade(&v);
    move || {
        let v = match glib::clone::Upgrade::upgrade(&v) {
            Some(val) => val,
            None => {
                glib::g_debug!(glib::CLONE_MACRO_LOG_DOMAIN, "Failed to upgrade v",);
                let ___return_value = || Bar { x: 0 };
                return ___return_value();
            }
        };
        {
            Bar { x: 1 }
        }
    }
}
```

And this is `wrapper`:

```
{
    let v = glib::clone::Downgrade::downgrade(&v);
    move |call: &(dyn Fn() -> Bar)| {
        let v = match glib::clone::Upgrade::upgrade(&v) {
            Some(val) => val,
            None => {
                panic!("failed to upgrade `v` (if you don't want to panic, use @default-return)",)
            }
        };
        {
            call()
        }
    }
}
```

Comment (sdroege) [reply glib/tests/clone.rs:467]: What's the problem here? I would expect it to return `Bar { x: 1 }` here and not `Bar { x: 0 }`. Why do you expect the other?

File: glib-macros/src/clone.rs [hunk 4]
```diff
@@ -381,7 +381,12 @@ fn get_expr(parts: &mut Peekable<ProcIter>) -> String {
                 }
                 ret.push_str(&p_s);
             }
-            Some(TokenTree::Group(g)) => ret.push_str(&group_to_string(g)),
+            Some(TokenTree::Group(g)) => {
+                ret.push_str(&group_to_string(g));
+                if total == 0 {
```
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:386]: I need to revert this change.

File: glib/tests/clone.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply glib/tests/clone.rs]: The problem was me. :)

**Comment by GuillaumeGomez** [comment]
Fixed it.

File: glib/tests/clone.rs [hunk 5]
```diff
@@ -451,29 +451,46 @@ fn test_clone_macro_typed_args() {
```
Comment (sdroege) [inline glib/tests/clone.rs:451]: Maybe the tests just above can also return something from their closures that can be tested?

**Review by sdroege — APPROVED** [review]

File: glib/tests/clone.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply glib/tests/clone.rs:451]: Another PR incoming then!
